File: java/104 ThreadPoolExecutor.java
package code;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Test {
    public static void main(String[] args) {
        ExecutorService service = Executors.newFixedThreadPool(5);

        for (int i = 0; i < 10; i++) {
            service.submit(new SleepTask());
        }

        service.shutdown();
    }

    static class SleepTask implements Runnable {
        @Override
        public void run() {
            Thread current = Thread.currentThread();
            System.out.printf("Started with Thread: %s%n", current.getName());
            try {
                Thread.sleep(getRandom() * 1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.printf("Ended with Thread: %s%n", current.getName());
        }

        private int getRandom() {
            return (int)(Math.random() * 5 + 1);
        }
    }
}

File: python/104 ThreadPoolExecutor.py
from concurrent.futures import ThreadPoolExecutor, wait, ALL_COMPLETED
import threading
import time
import random

class SleepTask:
    def __call__(self):
        thread_name = threading.current_thread().name
        print(f"Started with Thread: {thread_name}")
        time.sleep(random.randint(1, 5))
        print(f"Ended with Thread: {thread_name}")

executor = ThreadPoolExecutor(max_workers=5)
tasks = [executor.submit(SleepTask()) for _ in range(10)]

done, not_done = wait(tasks, timeout=10, return_when=ALL_COMPLETED)

if not_done:
    print("EMERGENCY SHUTDOWN")
    executor.shutdown(wait=False, cancel_futures=True)
else:
    executor.shutdown()
